lib/src/settings/embedder.dart

```diff
@@ -0,0 +1,262 @@
+import './distribution.dart';
+
+sealed class Embedder {
+  final String source;
+
+  const Embedder({
+    required this.source,
+  });
+
+  Map<String, Object?> toMap();
+
+  factory Embedder.fromMap(Map<String, Object?> map) {
+    final source = map['source'];
+
+    return switch (source) {
+      'openAi' => OpenAiEmbedder.fromMap(map),
+      'huggingFace' => HuggingFaceEmbedder.fromMap(map),
+      'userProvided' => UserProvidedEmbedder.fromMap(map),
+      'rest' => RestEmbedder.fromMap(map),
+      'ollama' => OllamaEmbedder.fromMap(map),
+      _ => throw Exception('unexpected embedder source'),
+    };
+  }
+}
+
+class OpenAiEmbedder extends Embedder {
+  final String? model;
+  final String? apiKey;
+  final String? documentTemplate;
+  final int? dimensions;
+  final Distribution? distribution;
+  final String? url;
+  final int? documentTemplateMaxBytes;
+  final bool? binaryQuantized;
+
+  const OpenAiEmbedder({
+    this.model,
+    this.apiKey,
+    this.documentTemplate,
+    this.dimensions,
+    this.distribution,
+    this.url,
+    this.documentTemplateMaxBytes,
+    this.binaryQuantized,
+  }) : super(source: 'openAi');
```

Comment:
for all embedders: we are typing the source twice, which might lead to typos, once here and once in `Embedder.fromMap`, we should probably unify this in a `static const sourceValue = 'openAi';`

Reply:
When I updated the code based on the comment above, the structure had already changed. The source is now typed only once without creating static variables. Let me know whether the current implementation works well for you

Reply:
I've updated the code, now it is typed only once